src/Core/AdminConsole/OrganizationFeatures/OrganizationUsers/ImportOrganizationUserCommand.cs

```diff
@@ -0,0 +1,305 @@
+    }
+
+    private async Task RemoveExistingExternalUsers(IEnumerable<string> removeUserExternalIds,
+            List<(OrganizationUserUserDetails ou, EventType e, DateTime? d)> events,
+            OrganizationUserImportData importUserData)
+    {
+        if (!removeUserExternalIds.Any())
+        {
+            return;
+        }
+
+        var existingUsersDict = importUserData.ExistingExternalUsers.ToDictionary(u => u.ExternalId);
+        var removeUsersSet = new HashSet<string>(removeUserExternalIds)
+            .Except(importUserData.NewUsersSet)
+            .Where(u => existingUsersDict.ContainsKey(u) && existingUsersDict[u].Type != OrganizationUserType.Owner)
+            .Select(u => existingUsersDict[u]);
+
+        await _organizationUserRepository.DeleteManyAsync(removeUsersSet.Select(u => u.Id));
+        events.AddRange(removeUsersSet.Select(u => (
+          u,
+          EventType.OrganizationUser_Removed,
+          (DateTime?)DateTime.UtcNow
+          ))
+        );
+    }
+
+    private async Task UpsertExistingUsers(IEnumerable<ImportedOrganizationUser> newUsers, OrganizationUserImportData importUserData)
+    {
+        if (!newUsers.Any())
+        {
+            return;
+        }
+
+        // Marry existing users
+        var existingUsersEmailsDict = importUserData.ExistingUsers
+            .Where(u => string.IsNullOrWhiteSpace(u.ExternalId))
+            .ToDictionary(u => u.Email);
+        var newUsersEmailsDict = newUsers.ToDictionary(u => u.Email);
+        var newAndExistingUsersIntersection = existingUsersEmailsDict.Keys.Intersect(newUsersEmailsDict.Keys).ToList();
+        var organizationUsers = (await _organizationUserRepository.GetManyAsync(importUserData.ExistingUsers.Select(u => u.Id).ToList())).ToDictionary(u => u.Id);
+        var usersToUpsert = new List<OrganizationUser>();
+
+        foreach (var user in newAndExistingUsersIntersection)
+        {
+            existingUsersEmailsDict.TryGetValue(user, out var existingUser);
+            organizationUsers.TryGetValue(existingUser.Id, out var organizationUser);
+
+            if (organizationUser != null)
+            {
+                organizationUser.ExternalId = newUsersEmailsDict[user].ExternalId;
+                usersToUpsert.Add(organizationUser);
+                importUserData.ExistingExternalUsersIdDict.Add(organizationUser.ExternalId, organizationUser.Id);
+            }
+        }
+        await _organizationUserRepository.UpsertManyAsync(usersToUpsert);
+    }
+
+    private async Task AddNewUsers(Organization organization,
+            IEnumerable<ImportedOrganizationUser> newUsers,
+            OrganizationUserImportData importUserData)
+    {
+        var userInvites = new List<OrganizationUserInviteCommandModel>();
+
+        foreach (var user in newUsers)
+        {
+            userInvites.Add(new OrganizationUserInviteCommandModel(user.Email, user.ExternalId));
+        }
```

Comment:
I don't think this matches existing logic - this is inviting all imported users, whereas the existing code excludes existing users: https://github.com/bitwarden/server/blob/66301bb4dc1ec7ea99d55301b14cfaeac216f1a4/src/Core/AdminConsole/Services/Implementations/OrganizationService.cs#L1172-L1175

Reply:
I think this is still an issue. ([context](https://github.com/bitwarden/server/pull/5800#discussion_r2128577928))